### DISCUSSION TITLE
Running tests with watch mode

### DISCUSSION THREAD

**Discussion by eyesight-pgs** [adonisjs/core#2860] [Help] [answered]
According to [this docs/article](https://docs.adonisjs.com/cookbooks/testing-adonisjs-apps#setup-for-adonisjs), we can run our tests with following command:
```bash
node -r @adonisjs/assembler/build/register japaFile.ts
```
I want to know how to run tests in watch mode, so that any source code changes runs the test again.

Thank you.

**Comment by arthurcvm** [comment]
Hi @eyesight-pgs, like as in docs, the Japa test runner up the own server to test, then you do not need start a server with `node ace serve --w`, only run `node -r @adonisjs/assembler/build/register japaFile.ts` command.

**Reply by eyesight-pgs** [reply]
Thanks for your reply.

I would like to know, if there is some way to watch for file changes.
Stand alone nodejs programs can be re-run with `nodemon`.
I would like to know if we can use japa tests `nodemon`.

**Reply by arthurcvm** [reply]
Oh, sorry, I understand now, you want run the Japa test like the jest watch. Japa test don't support this (I never found this), but you can use the [Husky](https://github.com/typicode/husky)  to run tests in pre-commit.

**Comment by thetutlage** [comment, marked as answer]
Use nodemon?

**Reply by eyesight-pgs** [reply]
I have created new AdonisJs project with
```bash
yarn create adonis-ts-app blog
```
Then followed [docs](https://docs.adonisjs.com/cookbooks/testing-adonisjs-apps#setup-for-adonisjs) for Japa test set up.
```
yarn add -D japa execa get-port
```
Created  `japaFile.ts` with given content
Created `test/example.spec.ts` file with content provided in docs.
Then ran test with node:
```
[es@espc blog]$ node -r @adonisjs/assembler/build/register japaFile.ts
{"level":30,"time":1624982081302,"pid":93684,"hostname":"espc","name":"blog","msg":"started server on 0.0.0.0:36089"}


Example
  ✓ assert sum (5ms)

 PASSED 

total         : 1
passed        : 1
duration      : 7ms
```
Everything is working.
---
Next I started testing with nodemon
I had nodemon installed globally
```
[es@espc blog]$ nodemon -v
2.0.7
```
Installed ts-node as dev dependency
```
yarn add -D ts-node
```
When I run the test with nodemon, I get error like `Type 'any' is not assignable to type 'never'`:
```
[es@espc blog]$ nodemon -r @adonisjs/assembler/build/register japaFile.ts
[nodemon] 2.0.7
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node -r @adonisjs/assembler/build/register japaFile.ts`

TSError: ⨯ Unable to compile TypeScript:
config/hash.ts:30:3 - error TS2322: Type 'any' is not assignable to type 'never'.

30   default: Env.get('HASH_DRIVER', 'argon'),
     ~~~~~~~

  node_modules/@adonisjs/hash/build/adonis-typings/hash.d.ts:101:9
    101         default: keyof HashersList;
                ~~~~~~~
    The expected type comes from property 'default' which is declared here on type 'HashConfig'



1 createTSError
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:587

2 reportTSError
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:591

3 getOutput
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:921

4 Object.compile
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:1189

5 Module.m._compile
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:1295

6 require.extensions.<computed>
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/ts-node/src/index.ts:1298

7 Object.newLoader [as .ts]
  /mnt/eyesight/testbed/adonis-testing-with-nodemon/blog/node_modules/pirates/lib/index.js:104



Example
  ✓ assert sum (2ms)

 PASSED 

total         : 1
passed        : 1
duration      : 4ms
[nodemon] clean exit - waiting for changes before restart
```
But test completed successfully. How to fix the errors? or should I ignore the errors?

**Reply by thetutlage** [reply]
Why have you installed `ts-node` when using `nodemon`?

**Reply by eyesight-pgs** [reply]
Because without installing ts-node I got following error
```
[es@espc blog]$ nodemon -r @adonisjs/assembler/build/register japaFile.ts
[nodemon] 2.0.7
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node -r @adonisjs/assembler/build/register japaFile.ts`
sh: line 1: ts-node: command not found
[nodemon] failed to start process, "ts-node" exec not found
[nodemon] Error
    at Bus.<anonymous> (/usr/lib/node_modules/nodemon/lib/nodemon.js:158:25)
    at Bus.emit (events.js:327:22)
    at ChildProcess.<anonymous> (/usr/lib/node_modules/nodemon/lib/monitor/run.js:184:11)
    at ChildProcess.emit (events.js:315:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
```

**Reply by thetutlage** [reply]
Yeah, but then the errors you are getting are from `ts-node`. Also, I think `nodemon` is making assumption to use tsnode because of the `.ts` file extension.

What you can do is, create the following file in the project root with the following contents ( Feel free to read the nodemon docs and add/modify other settings )

Filename: `nodemon.json`

```json
{
  "verbose": true,
  "execMap": {
    "ts": "node -r @adonisjs/assembler/build/register"
  }
}
```

The idea is to tell nodemon to use `node` with our `-r` flag and not ts-node for files ending with `.ts`

Also, I recommend install nodemon locally in your project and then create the following npm script to run tests.

```json
{
  "scripts": {
    "test": "nodemon japaFile.ts"
  }
}
```

**Reply by eyesight-pgs** [reply]
I followed your instructions and it worked perfectly. Thanks you so much :)
